### PR TITLE
Add tutorial 25

### LINKED ISSUE #204: Create an Advance Agent Tutorial
* Explain how to customize promp_parameter_resolver and final_answer_pattern

### PR DESCRIPTION
**Pull request by bilgeyucel**
Fixes #204

### CODE REVIEW

File: tutorials/25_Customizing_Agent.ipynb [hunk 1]
```diff
@@ -0,0 +1,4851 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "iHKX_ooAV1L2"
+   },
+   "source": [
+    "### 2) Initialize Memory\n",
+    "\n",
+    "Memory makes the interaction with the Agent more human-like with the chat ability. Between memory options in Haystack, `ConversationSummaryMemory` will create the summary of every three exchanges using [`philschmid/bart-large-cnn-samsum`](https://huggingface.co/philschmid/bart-large-cnn-samsum) model which is good at generating chat summaries.\n",
```
Comment (dfokina) [inline tutorials/25_Customizing_Agent.ipynb:443]:
```suggestion
    "Memory makes the interaction with the Agent more human-like. Haystack offers various memory options, one of which, `ConversationSummaryMemory`, will create the summary of every three exchanges. You will be using the [`philschmid/bart-large-cnn-samsum`](https://huggingface.co/philschmid/bart-large-cnn-samsum) model which is especially good at generating chat summaries.\n",
```

File: tutorials/25_Customizing_Agent.ipynb [hunk 2]
```diff
@@ -0,0 +1,4851 @@
...
+    "\n",
+    "> To discover other memory options in Haystack, check out [Good Listener: How Memory Enables Conversational Agents](https://haystack.deepset.ai/blog/memory-conversational-agents#summarizing-memory)."
```
Comment (dfokina) [inline tutorials/25_Customizing_Agent.ipynb:445]:
```suggestion
    "> To discover other memory options in Haystack, check out [Good Listener: How Memory Enables Conversational Agents](https://haystack.deepset.ai/blog/memory-conversational-agents#summarizing-memory) blog post."
```

File: tutorials/25_Customizing_Agent.ipynb [hunk 3]
```diff
@@ -0,0 +1,4851 @@
...
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "1RPZ-ccXXUSN",
+    "outputId": "dae063d4-1e04-4656-d384-cc34f3c93830"
+   },
+   "outputs": [],
+   "source": [
+    "conversational_agent.run(\"How did Rhodes Statue look like?\")"
```
Comment (dfokina) [inline tutorials/25_Customizing_Agent.ipynb:667]:
```suggestion
    "conversational_agent.run(\"What did Rhodes Statue look like?\")"
```

**Review by dfokina — COMMENTED** [review]
Leaving some suggestion on wording changes :)

File: tutorials/25_Customizing_Agent.ipynb [hunk 4]
```diff
@@ -0,0 +1,4851 @@
+{
```
Comment (TuanaCelik) [inline tutorials/25_Customizing_Agent.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.            "transcript": agent_step.transcript,
</pre><p>Where does agent_step come from? and also then agent_step.transcript? Are resolver functions always supposed to take in these three 'query, agent and agent_step'? If yes, I would rephrase the top to:</p><ol><li>Say this exactly, a resolver function should always take in these 3 arguments</li><li>Describe what AgentStep is or link to the documentation page</li></ol><p>Further, can someone please explain why we need this? Is it optional? Or is it always needed? Are there no default resolver functions? Most agents will always have query, tool_names_withn_descriptions, transcript and memory right? </p><p>Are we only explaining this as a optional customization? If yes, perfect, but as a reader I may think it's required. So the only change I would suggest is to mention that this is something that has a default setup, but this is how you customize it :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-tutorials/pull/223/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

File: tutorials/25_Customizing_Agent.ipynb [hunk 5]
```diff
@@ -0,0 +1,4851 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "a5aqoO2wUFk-"
+   },
+   "source": [
+    "### 4) Define a Tool with the Generative QA Pipeline\n",
+    "\n",
+    "Later, you will provide this Tool to your Agent."
```
[the remark this answers is not in the document]
Comment (bilgeyucel) [reply tutorials/25_Customizing_Agent.ipynb:359]: I wanted to mention that "You're creating this Tool to pass to your Agent. So, define a Tool with the Generative QA Pipeline and later you will provide this Tool to your Agent."

File: tutorials/25_Customizing_Agent.ipynb (hunk 4 above)
Comment (bilgeyucel) [reply tutorials/25_Customizing_Agent.ipynb:1]: As we discussed, we will add AgentStep to the documentation. I also added more information to the tutorial to clarify the points you mentioned 👍

**Comment by bilgeyucel** [comment]
Hi @dfokina and @TuanaCelik, I believe I resolved all your comments. Let me know if you have more questions and suggestions 😊

File: tutorials/25_Customizing_Agent.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (dfokina) [reply tutorials/25_Customizing_Agent.ipynb:359]: Ah, okay! Let's leave it like it is then!

**Review by dfokina — APPROVED** [review]

**Comment by dfokina** [comment]
LGTM! 😊

File: tutorials/25_Customizing_Agent.ipynb [hunk 6]
```diff
@@ -0,0 +1,4854 @@
+{
```
Comment (TuanaCelik) [inline tutorials/25_Customizing_Agent.ipynb:1]: <p>I would have the 'Keep in mind' note at the top and also add: 'here we show an example of how you might create your own resolver for your custom usecase. Also, if they are defineable, maybe list some example use cases that we already have resolvers for.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-tutorials/pull/223/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>
